### PR TITLE
Resolver error formatting

### PR DESCRIPTION
**Pull request by segiddins**
Closes https://github.com/bundler/bundler/issues/3803

### CODE REVIEW

**Comment by segiddins** [comment]
We should remember to apply the following diff once https://github.com/CocoaPods/Molinillo/pull/24 is merged and released:

```diff
diff --git a/lib/bundler/resolver.rb b/lib/bundler/resolver.rb
index f21c9cf..ec3b2a4 100644
--- a/lib/bundler/resolver.rb
+++ b/lib/bundler/resolver.rb
@@ -35,6 +35,9 @@ module Bundler
             depth = 2
             tree.each do |req|
               t << "  " * depth << %(#{clean_req req})
+              if spec = conflict.activated_by_name[req.name]
+                t << %(, resolved to #{spec.version},)
+              end
               t << %( depends on) unless tree.last == req
               t << %(\n)
               depth += 1
```

**Comment by segiddins** [comment]
@indirect this is now ready for review

File: spec/install/gems/flex_spec.rb [hunk 1]
```diff
@@ -200,11 +200,11 @@
             rack (= 0.9.1)
 
           In Gemfile:
-            rack_middleware (>= 0) ruby depends on
-              rack (= 0.9.1) ruby
+            rack-obama (= 2.0), resolved to 2.0, depends on
+              rack (= 1.2)
```
Comment (indirect) [inline spec/install/gems/flex_spec.rb]: Now that I'm reading it, I'd prefer `rack-obama (= 2.0) was resolved to 2.0, which depends on\n  rack(= 1.2)`. Sorry to nitpick on you. :stuck_out_tongue_closed_eyes:

**Comment by indirect** [comment]
This is a huuuuuge improvement in the UX of requirement conflicts. :clap: :raised_hands:

**Comment by segiddins** [comment]
@indirect done

**Comment by indirect** [comment]
Woo! @homu r+

**Comment by indirect** [comment]
@segiddins looks like it's broken on 1.8.7

**Comment by segiddins** [comment]
Looks like something has an unstable comparable implementation to me?

**Comment by indirect** [comment]
Hashes literally do not have a stable iteration order in 1.8.7; you'll need to sort the hash before printing the conflict message or store conflicts in an array of tuples.

**Comment by segiddins** [comment]
@indirect should work now!

**Comment by indirect** [comment]
@homu r+

**Comment by segiddins** [comment]
@indirect want to give me a hand on this?

**Comment by segiddins** [comment]
@indirect should _actually_ be fixed now

**Comment by indirect** [comment]
Heh. Thanks!

**Comment by indirect** [comment]
@homu r+

**Comment by segiddins** [comment]
@homu r+

File: lib/bundler/dep_proxy.rb [hunk 2]
```diff
@@ -29,13 +29,16 @@ def requirement
     end
 
     def to_s
-      "#{name} (#{requirement}) #{__platform}"
+      s = name.dup
+      s << " (#{requirement})" unless requirement == Gem::Requirement.default
```
Comment (indirect) [inline lib/bundler/dep_proxy.rb]: Looks like Gem::Requirement.default doesn't exist on RubyGems 1.5 and lower, so we're adding "(runtime)" there? :/

File: lib/bundler/resolver.rb [hunk 3]
```diff
@@ -1,51 +1,40 @@
...
-
       def message
-        conflicts.values.flatten.reduce("") do |o, conflict|
-          o << %(Bundler could not find compatible versions for gem "#{conflict.requirement.name}":\n)
+        conflicts.sort.reduce("") do |o, (name, conflict)|
+          o << %(Bundler could not find compatible versions for gem "#{name}":\n)
           if conflict.locked_requirement
             o << %(  In snapshot (#{Bundler.default_lockfile.basename}):\n)
-            o << %(    #{clean_req conflict.locked_requirement}\n)
+            o << %(    #{conflict.locked_requirement}\n)
```
Comment (indirect) [inline lib/bundler/resolver.rb]: This is actually a `Gem::Dependency`, whose to_s method includes "runtime" on RubyGems 1.5 and older. How is it not a `DepProxy`? No one knows...

**Comment by segiddins** [comment]
@homu r+

**Comment by segiddins** [comment]
Wooooo! Finally! Thanks for the help on this one, @indirect !

**Comment by indirect** [comment]
ugh, just realized that I should be reading the platform from the dep... the output will be wrong if the platform isn't Raunt. :P

**Comment by segiddins** [comment]
mhmm :P

**Comment by segiddins** [comment]
Actually, @indirect, do we even have the platform for locked gems?

**Comment by indirect** [comment]
no idea! that's... kind of distressing, actually.

**Comment by indirect** [comment]
Opened #3906 to print platform(s) correctly.

**Pull request merged**